# Homework Assignment 3: Data Formats and Access

## Part 1

In [1]:
import requests
import json

In [2]:
import json
import requests
# Here's a sample JSON structure similar to what APIs return
sample_json = '''
{
  "station": "USC00305800",
  "name": "New York Central Park",
  "location": {
    "latitude": 40.7789,
    "longitude": -73.9692
  },
  "observations": [
    {"date": "2023-01-01", "temperature": 32, "precipitation": 0.0},
    {"date": "2023-01-02", "temperature": 28, "precipitation": 0.5},
    {"date": "2023-01-03", "temperature": 35, "precipitation": 0.0},
    {"date": "2023-01-04", "temperature": 38, "precipitation": 0.2},
    {"date": "2023-01-05", "temperature": 41, "precipitation": 0.0}
  ]
}
'''

# Parse the JSON
data = json.loads(sample_json)

# Access nested data
print("Station:", data['station'])
print("Location:", data['location'])
print("First observation:", data['observations'][0])

Station: USC00305800
Location: {'latitude': 40.7789, 'longitude': -73.9692}
First observation: {'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}


In [3]:
# 1. Extract and print all dates and temperatures (8 points)
print("Date, Temperature")
for obs in data['observations']:
    print(f"{obs['date']}, {obs['temperature']}")

Date, Temperature
2023-01-01, 32
2023-01-02, 28
2023-01-03, 35
2023-01-04, 38
2023-01-05, 41


In [6]:
# 2. Calculate average temperature (8 points)
total_temp = 0
count = 0
for obs in data['observations']:
    total_temp += obs['temperature']
    count += 1

avg_temp = total_temp / count
print(f"Average temperature: {avg_temp}°F")

Average temperature: 34.8°F


In [8]:
# 3. Find days with precipitation (9 points)
print("\nDays with precipitation:")
for obs in data['observations']:
    if obs['precipitation'] > 0:
        print(f"{obs['date']}: {obs['precipitation']} inches")


Days with precipitation:
2023-01-02: 0.5 inches
2023-01-04: 0.2 inches


## Part 2

In [20]:
import pooch

# Set up Pooch to download a file
# This example downloads a small air quality dataset
file_path = pooch.retrieve(
    url="https://github.com/pandas-dev/pandas/raw/main/doc/data/air_quality_no2.csv",
    known_hash=None)

In [21]:
print("File downloaded to:", file_path)
print("File exists:", os.path.exists(file_path))

File downloaded to: /home/ank2190/.cache/pooch/458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
File exists: True


In [22]:
import os

# 1. Verify the file was downloaded (5 points)
# Check the file size
file_size = os.path.getsize(file_path)
print(f"File size: {file_size} bytes")

File size: 31984 bytes


In [24]:
# YOUR CODE HERE: open the file and count how many lines it has
line_count = 0 

In [25]:
print(f"Number of line: {line_count}")

Number of line: 0


In [ ]:
import pooch 

file_path = pooch.retrieve(
    url="https://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods",
    known_hash=None)

In [ ]:
file_size = os.path.getsize(file_path)
print(f"File size: {file_size} bytes")

In [87]:
print("File downloaded to:", file_path)
print("File exists:", os.path.exists(file_path))

File downloaded to: /home/ank2190/.cache/pooch/60b010059f33d2c570306c69121de5d7-USW00023183.csv
File exists: True


In [88]:
print(f"Number of lines: {line_count}")

Number of lines: 0


In [51]:
# YOUR CODE HERE: add your file from task 2

In [89]:
import pooch
import os

In [90]:
cache_dir = pooch.os_cache("pooch")
print(f"Cache directory: {cache_dir}")

Cache directory: /home/ank2190/.cache/pooch


In [91]:
files = os.listdir(cache_dir)
print("\nDownloaded files:")
for file in files:
    print(f" - {file}")


Downloaded files:
 - 458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
 - 52fd1a232f8b1cfa99e8f431fdc297d6-global-ocean-heat-content
 - ed6975dbcf2dfea3336051f0be8fa2a8-global-marine
 - b3fc8e966c342139ca6050eb6cac56f8-global-summary-of-the-year
 - 60b010059f33d2c570306c69121de5d7-USW00023183.csv


In [77]:
print("\nDownlodedfiles:")
for file in files:
    print(type(data))


Downlodedfiles:
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>


In [78]:
print(data)

{'station': 'USC00305800', 'name': 'New York Central Park', 'location': {'latitude': 40.7789, 'longitude': -73.9692}, 'observations': [{'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}, {'date': '2023-01-02', 'temperature': 28, 'precipitation': 0.5}, {'date': '2023-01-03', 'temperature': 35, 'precipitation': 0.0}, {'date': '2023-01-04', 'temperature': 38, 'precipitation': 0.2}, {'date': '2023-01-05', 'temperature': 41, 'precipitation': 0.0}]}


In [92]:
# 3. Create a data inventory (5 points)
# List all the files you've downloaded in this assignment
print("\nData Inventory:")
print("1. air_quality_no2.csv - Air quality NO2 measurements")
print("2. phoeinix_presipitation.csv - Phoenix precipitation 1934-2025")


Data Inventory:
1. air_quality_no2.csv - Air quality NO2 measurements
2. phoeinix_presipitation.csv - Phoenix precipitation 1934-2025


In [ ]:
# List all variables and their types
for var_name in dir():
    if not var_name.startswith('_'): 
        var_value = eval(var_name)
        print(f"{var_name}: {type(var_value)}")

## Part 3

In [93]:
import requests

# OPeNDAP provides metadata in different formats
# We'll get basic info about a climate dataset

base_url = "http://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"

# Get DDS (Dataset Descriptor Structure) - describes the structure
dds_url = base_url + ".dds"
response = requests.get(dds_url)

print("Dataset Structure:")
print(response.text[:500])  # Print first 500 characters

Dataset Structure:
Dataset {
    Float32 T[T = 324];
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



In [103]:
import requests 

base_url = "https://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"

dds_url = base_url + ".dds"
response = requests.get(dds_url)
print("DDS OUTPUT:")
print(response.text)

DDS OUTPUT:
Dataset {
    Float32 T[T = 324];
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



### 1. Dimensions and Variables

**Dimension names:** 
- T (time) = 324
- Y (latitude) = 360
- X (longitude) = 720

**Main variable name:**
precipitation (rain)

In [106]:
das_url = base_url + ".das"

response = requests.get(das_url)
print("DAS OUTPUT (first 1000 characters):")
print(response.text[:1000])

print("\n\nFULL DAS OUTPUT:")
print(response.text)

DAS OUTPUT (first 1000 characters):
Attributes {
    T {
        Float32 pointwidth 1.0;
        String calendar "360";
        Int32 gridtype 0;
        String units "months since 1960-01-01";
    }
    Y {
        Float32 pointwidth 0.5;
        String standard_name "latitude";
        Int32 gridtype 0;
        String units "degree_north";
    }
    X {
        Float32 pointwidth 0.5;
        String standard_name "longitude";
        Int32 gridtype 1;
        String units "degree_east";
    }
    rain {
        Int32 pointwidth 0;
        String standard_name "lwe_precipitation_rate";
        Float32 file_missing_value -999.0;
        String units "mm/day";
        String history "Boxes with less than 0.0% dropped";
        Float32 missing_value NaN;
        String long_name "Monthly Precipitation";
    }
NC_GLOBAL {
    String Conventions "IRIDL";
}
}



FULL DAS OUTPUT:
Attributes {
    T {
        Float32 pointwidth 1.0;
        String calendar "360";
        Int32 gridtype 0;
   

### 3. Dataset Info

**What does this dataset contain?:**
Monthly precipitation data (gauge-based rainfall measurements from NOAA NCEP CPC Unified Precipitation dataset)

**Time period covered:**
Jan 1960 to present (324 months from 1960-01-01, ~27 years of data)

**Geographic region:**
Global coverage
Latitude: 360 grid points (0.5° resolution) = -90° to 90° (pole to pole)
Longitude: 720 grid points (0.5° resolution) = -180° to 180° (global)

**Units of main variable:**
mm/day (millimeters per day)